Blog/using multiple same components (#226)
infinum/eightshift-docs: website/blog/2023-06-12-multiple-same-components.md

While at first glance it seems strange to have this naming scheme, it actually allows us to figure out the hierarchy just by looking at the attribute name.

## A step-by-step example

The block we'll be creating as an example will be a block which we can use for some sort of comparison or listing pros and cons. To create it, we will need the following:
- two `Heading` components
- two `List` components

The easiest way to create a new block is by using the boilerplate command

`wp boilerplate blocks use-block --name=example`

Once the new block is added to your project, rename it. Also, don’t forget to update all file names and imports in JS.

We should start with `manifest.json`, where we define the components we'll use and set the default attributes. For now, just define the components and their keys:

```json
"components": {
	"heading": "heading",
	"lists": "lists",
	"secondaryHeading": "heading",
	"secondaryLists": "lists"
}
```

The first section has the default key names, while the "duplicates" have different key names.

> After manifest, we can move to the JS part. As this is a fairly simple block without any advanced options or layouts, we need to add two `HeadingEditor` components and two `ListEditor` components. To make styling easier, we can separate them in two `div` elements. When you’re finished, your code should look like this:

Reviewer: ```suggestion
After manifest, we can move to the JS part. As this is a fairly simple block without any advanced options or layouts, we need to add two `HeadingEditor` components and two `ListsEditor` components. To make styling easier, we can separate them in two `div` elements. When you’re finished, your code should look like this:
```

Author: Oh, great catch! Tnx!

---
title: Using multiple same components
description: Explains how to use multiple same components inside
slug: multiple-same-components
authors: obradovic

date: 2023-06-12
tags: [eightshift, boilerplate, components, blocks]
hide_table_of_contents: false
---

> From time to time, you may have to create a block or a more complex component that is using the same component two or more times. In this blog post, we'll explain how it works in a bit more detail.

Reviewer 2: ```suggestion
From time to time, you may need to create a block or a more complex component that uses more than one instance the same component. In this blog post, we'll explain how it works in a bit more detail.
```

## Manifest and attributes
The way multiple same components work is by having a different key. Here is an example of the Card component:

```json
"components": {
	"image": "image",
	"intro": "heading",
	"heading": "heading",
	"paragraph": "paragraph",
	"button": "button"
},
```
As you can see, one heading component has the `intro` key, while the other one has the `heading` key. That way the **intro** heading component is being referred to as `intro` so there is no mixup with the attribute values between the two heading blocks. This can be seen when setting the default attributes:

```json
"cardIntroSize": {
	"type": "string",
	"default": "tiny"
},
"cardHeadingSize": {
	"type": "string",
	"default": "big"
}
```

## The "props" Helper
This method does all the heavy lifting for us. It replaces the default attribute names with the ones we provide. There is both the PHP and JS version of it.

```php

> Components::render(
> 	'heading',
> 	Components::props('intro', $attributes, [
> 		'selectorClass' => 'intro',
> 		'blockClass' => $componentClass
> 	])
> ),

Reviewer 2: ```suggestion
Components::render('heading', Components::props('intro', $attributes, [
	'selectorClass' => 'intro',
	'blockClass' => $componentClass
]))
```
Just a suggestion to be more consistent with the props helper inside (first argument on the same line as the function call).

If you like how the existing one looks, feel free to ignore, but then just remove the comma at the end 😄

```

```js

<HeadingEditor

> {...props('intro', attributes, {

Reviewer 2: Looks like indentation is a bit wonky here 😄

> <HeadingEditor
> {...props('intro', attributes, {
> selectorClass: 'intro',
> blockClass: componentClass,
> })}
> />

Reviewer 2: ```suggestion
<HeadingEditor
	{...props('intro', attributes, {
		selectorClass: 'intro',
		blockClass: componentClass,
	})}
/>
```

`cardCardHeading` - first is the Card block name, then the Card component name and finally Heading component name.

The Heading component that is called Intro then has the prefix `cardCardIntro`. This prefix is then added to the attribute name, which then finally results in `cardCardHeadingSize` and `cardCardIntroSize` attributes, for example.

While at first glance it seems strange to have this naming scheme, it actually allows us to figure out the hierarchy just by looking at the attribute name.

## A step-by-step example

The block we'll be creating as an example will be a block which we can use for some sort of comparison or listing pros and cons. To create it, we will need the following:
- two `Heading` components
- two `List` components

The easiest way to create a new block is by using the boilerplate command

`wp boilerplate blocks use-block --name=example`

Once the new block is added to your project, rename it. Also, don’t forget to update all file names and imports in JS.

We should start with `manifest.json`, where we define the components we'll use and set the default attributes. For now, just define the components and their keys:

```json

> 	"components": {
> 		"heading": "heading",
> 		"lists": "lists",
> 		"secondaryHeading": "heading",
> 		"secondaryLists": "lists"
> 	}

Reviewer 2: ```suggestion
"components": {
	"heading": "heading",
	"lists": "lists",
	"secondaryHeading": "heading",
	"secondaryLists": "lists"
}
```

```

The first section has the default key names, while the "duplicates" have different key names.

After manifest, we can move to the JS part. As this is a fairly simple block without any advanced options or layouts, we need to add two `HeadingEditor` components and two `ListEditor` components. To make styling easier, we can separate them in two `div` elements. When you’re finished, your code should look like this:

```jsx
import React from 'react';
import { __ } from '@wordpress/i18n';
import { HeadingEditor } from '../../../components/heading/components/heading-editor';
import { ListsEditor } from '../../../components/lists/components/lists-editor';

import { props } from '@eightshift/frontend-libs/scripts';

export const ComparisonEditor = ({ attributes, setAttributes }) => {
	const {
		blockClass,
	} = attributes;

	return (
		<div className={blockClass}>

> 			<div className={`${blockClass}__primary`}>

Reviewer 2: This should use the `selector` helper 😄

Author: I remember there was a discussion a while back that there is no need to use the `selector` helper if the class name is simple as in the example above.

But if recently it was decided to use `selector` helper always, even in these cases, I can update the code accordingly, no problem 😄

				<HeadingEditor
					{...props('heading', attributes, {
						blockClass: blockClass,
						setAttributes: setAttributes,
					})}
					/>

				<ListsEditor
					{...props('lists', attributes, {
						blockClass: blockClass,
						setAttributes: setAttributes,
					})}
				/>
			</div>

> 			<div className={`${blockClass}__secondary`}>

Reviewer 2: Same as 🆙

			<ListsOptions
				{...props('lists', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				reducedBottomSpacing
			/>

			<HeadingOptions
				{...props('secondaryHeading', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				label={__('Secondary Heading', 'eightshift')}
				reducedBottomSpacing
			/>

			<ListsOptions
				{...props('secondaryLists', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				label={__('Secondary Lists', 'eightshift')}
				reducedBottomSpacing
			/>
		</PanelBody>
	);
};
```

Adding the PHP part should be simple, but here is the code for reference:

```php
<?php

/**
 * Template for the Comparison Block view.
 *
 * @package Eightshift
 */

use EightshiftVendor\EightshiftLibs\Helpers\Components;

$globalManifest = Components::getManifest(dirname(__DIR__, 2));
$manifest = Components::getManifest(__DIR__);

$blockClass = $attributes['blockClass'] ?? '';

$unique = Components::getUnique();

?>

<div class="<?php echo esc_attr($blockClass); ?>">
	<?php echo Components::outputCssVariables($attributes, $manifest, $unique, $globalManifest); ?>

> 	<div class="<?php echo esc_attr("{$blockClass}__primary"); ?>">

Reviewer 2: Same as 🆙

			Components::render('lists', Components::props('lists', $attributes));
		?>
	</div>
	<div class="<?php echo esc_attr("{$blockClass}__secondary"); ?>">
		<?php
			echo Components::render('heading', Components::props('secondaryHeading', $attributes)),
			Components::render('lists', Components::props('secondaryLists', $attributes));
		?>
	</div>
</div>
```

The block should work properly now, but the two lists look the same. To make a difference between the two, we can change the default list colors in the manifest. If we don’t have the colors we want already available in the project, first we need to add them to the global manifest.

For a detailed explanation on how to add new colors to your project, you can read a previous blog post about [modifying blocks](/blog/modifying-blocks-color-theme#adding-new-colors-to-your-project).

Now we have to add these new colors to the Lists component `manifest.json`. In `options` key, find the `listsColor` and add your new colors.

When the new colors are added to the Lists component, we can set these new colors as defaults by adding the following attributes in `manifest.json` of our Comparison block:

```json

> "comparisonListsColor": {
> 			"type": "string",
> 			"default": "green-haze"
> 		},
> 		"comparisonSecondaryListsColor": {
> 			"type": "string",
> 			"default": "milano-red"
> 		}

Reviewer 2: ```suggestion
"comparisonListsColor": {
	"type": "string",
	"default": "green-haze"
},
"comparisonSecondaryListsColor": {
	"type": "string",
	"default": "milano-red"
}
```

The easiest way to create a new block is by using the boilerplate command

`wp boilerplate blocks use-block --name=example`

Once the new block is added to your project, rename it. Also, don’t forget to update all file names and imports in JS.

We should start with `manifest.json`, where we define the components we'll use and set the default attributes. For now, just define the components and their keys:

```json
"components": {
	"heading": "heading",
	"lists": "lists",
	"secondaryHeading": "heading",
	"secondaryLists": "lists"
}
```

The first section has the default key names, while the "duplicates" have different key names.

After manifest, we can move to the JS part. As this is a fairly simple block without any advanced options or layouts, we need to add two `HeadingEditor` components and two `ListsEditor` components. To make styling easier, we can separate them in two `div` elements. When you’re finished, your code should look like this:

```jsx
import React from 'react';
import { __ } from '@wordpress/i18n';
import { HeadingEditor } from '../../../components/heading/components/heading-editor';
import { ListsEditor } from '../../../components/lists/components/lists-editor';

> import { props, classnames, selector } from '@eightshift/frontend-libs/scripts';

Reviewer 2: ```suggestion
import { props, selector } from '@eightshift/frontend-libs/scripts';
```

We should start with `manifest.json`, where we define the components we'll use and set the default attributes. For now, just define the components and their keys:

```json
"components": {
	"heading": "heading",
	"lists": "lists",
	"secondaryHeading": "heading",
	"secondaryLists": "lists"
}
```

The first section has the default key names, while the "duplicates" have different key names.

After manifest, we can move to the JS part. As this is a fairly simple block without any advanced options or layouts, we need to add two `HeadingEditor` components and two `ListsEditor` components. To make styling easier, we can separate them in two `div` elements. When you’re finished, your code should look like this:

```jsx
import React from 'react';
import { __ } from '@wordpress/i18n';
import { HeadingEditor } from '../../../components/heading/components/heading-editor';
import { ListsEditor } from '../../../components/lists/components/lists-editor';

import { props, classnames, selector } from '@eightshift/frontend-libs/scripts';

export const ComparisonEditor = ({ attributes, setAttributes }) => {
	const {
		blockClass,
	} = attributes;

> 	const comparisonPrimaryClass = classnames(
> 		selector(blockClass, blockClass, 'primary'),
> 	);

Reviewer 2: ```suggestion
	const comparisonPrimaryClass = selector(blockClass, blockClass, 'primary');
```
Don't need `classnames` if using only one selector 😄

Reviewer: > Couple more 😄
> 
> Regarding the comment on (not) using the `selector` helper, honestly cannot remember the discussion, but it would feel a bit weird to have a mix of using and not using it. Of course, if others feel like no `selector` is the way to, I'm totally fine with that option.

I feel that the verbose method is better just for sake of a cleaner git history in case something gets added later on, just sayinn 😄

> 	const comparisonSecondaryClass = classnames(
> 		selector(blockClass, blockClass, 'secondary'),
> 	);

Reviewer 2: ```suggestion
	const comparisonSecondaryClass = selector(blockClass, blockClass, 'secondary');
```

	return (
		<div className={blockClass}>
			<div className={comparisonPrimaryClass}>
				<HeadingEditor
					{...props('heading', attributes, {
						blockClass: blockClass,
						setAttributes: setAttributes,
					})}

> 					/>

Reviewer 2: Please check the indent here

Author: Fixed it, thanks for pointing it out 😄 sometimes when copying the indent gets messed up and I totally miss it

			<HeadingOptions
				{...props('heading', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				reducedBottomSpacing
			/>

			<ListsOptions
				{...props('lists', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				reducedBottomSpacing
			/>

			<HeadingOptions
				{...props('secondaryHeading', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				label={__('Secondary Heading', 'eightshift')}
				reducedBottomSpacing
			/>

			<ListsOptions
				{...props('secondaryLists', attributes, {
					setAttributes,
					options: getOptions(attributes, manifest)
				})}
				label={__('Secondary Lists', 'eightshift')}
				reducedBottomSpacing
			/>
		</PanelBody>
	);
};
```

Adding the PHP part should be simple, but here is the code for reference:

```php
<?php

/**
 * Template for the Comparison Block view.
 *
 * @package Eightshift
 */

use EightshiftVendor\EightshiftLibs\Helpers\Components;

$globalManifest = Components::getManifest(dirname(__DIR__, 2));
$manifest = Components::getManifest(__DIR__);

$blockClass = $attributes['blockClass'] ?? '';

> $comparisonPrimaryClass = Components::classnames([
> 	Components::selector($blockClass, $blockClass, 'primary'),
> ]);

Reviewer 2: ```suggestion
$comparisonPrimaryClass = Components::selector($blockClass, $blockClass, 'primary');
```

> $comparisonSecondaryClass = Components::classnames([
> 	Components::selector($blockClass, $blockClass, 'secondary'),
> ]);

Reviewer 2: ```suggestion
$comparisonSecondaryClass = Components::selector($blockClass, $blockClass, 'secondary');
```
